pkg/controllers/provisioningv2/rke2/rkecontrolplane/controlplane.go

```diff
@@ -85,7 +86,15 @@ func (h *handler) OnChange(obj *rkev1.RKEControlPlane, status rkev1.RKEControlPl
 func (h *handler) OnRemove(_ string, cp *rkev1.RKEControlPlane) (*rkev1.RKEControlPlane, error) {
 	status := cp.Status
 	cp = cp.DeepCopy()
-	err := rke2.DoRemoveAndUpdateStatus(cp, h.doRemove(cp), h.rkeControlPlaneController.EnqueueAfter)
+
+	provCluster, err := h.provClusterCache.Get(cp.Namespace, cp.Name)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return cp, err
+	}
+
+	// only delete machines from custom clusters, node driver will clean up themselves
+	deleteMachines := len(provCluster.Spec.RKEConfig.MachinePools) == 0
+	err = rke2.DoRemoveAndUpdateStatus(cp, h.doRemove(cp, deleteMachines), h.rkeControlPlaneController.EnqueueAfter)
```

Comment:
Shouldn't we address the err here if its non nil? If we were unable to update the cp Status perhaps we could try again or return?

Suggestion

```
if err != nil || equality.Semantic.DeepEqual(status, cp.Status) {
  return cp, err
}
```

Reply:
Not quite, because we want to update the status of the controlplane object to expose the machine deletion status in the UI, which we may never do if we always return an error. `DoRemoveAndUpdateStatus` doesn't actually update status, it just populates the expected conditions, which is a confusing name and IMO bad paradigm, but that is something to be fixed next release cycle IMO.